File: runtimes/openai-agents/src/azureclaw_runtime_openai_agents/otel.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""
OpenTelemetry auto-init for the OpenAI Agents adapter.

The router sidecar exposes an OTLP/HTTP collector at
`/v1/traces` and `/v1/metrics`. We default to the loopback address so
egress-guard can keep the pod sealed; an operator may override via
`OTEL_EXPORTER_OTLP_ENDPOINT` if they front the collector elsewhere.

`init_telemetry()` is idempotent: a second call replaces nothing — once
a tracer/meter provider is set globally we leave it alone. This keeps
re-imports during test runs from blowing up the global state.
"""

from __future__ import annotations

import logging
import os
import threading
from typing import Optional

logger = logging.getLogger(__name__)

DEFAULT_OTLP_ENDPOINT = "http://127.0.0.1:8443/v1/traces"
DEFAULT_OTLP_METRICS_ENDPOINT = "http://127.0.0.1:8443/v1/metrics"

_init_lock = threading.Lock()
_initialized = False


def _is_initialized() -> bool:
    return _initialized


def _otlp_endpoint(env_var: str, default: str) -> str:
    raw = os.environ.get(env_var)
    return raw if raw else default


def init_telemetry(
    service_name: str,
    service_version: str = "0.1.0",
    *,
    traces_endpoint: Optional[str] = None,
    metrics_endpoint: Optional[str] = None,
) -> None:
    """Configure global OTel tracer + meter providers.

    Calling this more than once is a no-op (subsequent calls log and
    return). `traces_endpoint`/`metrics_endpoint` override the
    `OTEL_EXPORTER_OTLP_*` env vars when given (used by tests).
    """
    global _initialized
    with _init_lock:
        if _initialized:
            logger.debug("init_telemetry: already initialized, skipping")
            return

        # Imports deferred so the module imports cheaply when telemetry
        # is disabled (e.g. unit tests that just probe constants).
        from opentelemetry import metrics, trace
        from opentelemetry.exporter.otlp.proto.http.metric_exporter import (
            OTLPMetricExporter,
        )
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )
        from opentelemetry.sdk.metrics import MeterProvider
        from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        traces_url = traces_endpoint or _otlp_endpoint(
            "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT",
            _otlp_endpoint("OTEL_EXPORTER_OTLP_ENDPOINT", DEFAULT_OTLP_ENDPOINT),
        )
        metrics_url = metrics_endpoint or _otlp_endpoint(
            "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT",
            DEFAULT_OTLP_METRICS_ENDPOINT,
        )

        resource = Resource.create(
            {
                "service.name": service_name,
                "service.version": service_version,
                "service.namespace": "azureclaw",
                "azureclaw.runtime.kind": "OpenAIAgents",
            }
        )

        tracer_provider = TracerProvider(resource=resource)
        tracer_provider.add_span_processor(
            BatchSpanProcessor(OTLPSpanExporter(endpoint=traces_url))
        )
        trace.set_tracer_provider(tracer_provider)

        metric_reader = PeriodicExportingMetricReader(
            OTLPMetricExporter(endpoint=metrics_url),
        )
        meter_provider = MeterProvider(
            resource=resource, metric_readers=[metric_reader]
        )
        metrics.set_meter_provider(meter_provider)

        _instrument_httpx()

        _initialized = True
        logger.info(
            "azureclaw OTel initialized: service=%s traces=%s metrics=%s",
            service_name,
            traces_url,
            metrics_url,
        )


def _instrument_httpx() -> None:
    """Auto-instrument httpx so every LLM/tool HTTP call is a span.

    The OpenAI Python SDK uses httpx under the hood, so this transitively
    captures every model call. Failures are downgraded to warnings —
    telemetry must never crash the agent.
    """
    try:
        from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor

        HTTPXClientInstrumentor().instrument()
    except Exception as exc:  # pragma: no cover - defensive
        logger.warning("httpx instrumentation failed: %s", exc)


def _reset_for_tests() -> None:
    """Test-only hook to reset the module-level init flag."""
    global _initialized
    with _init_lock:
        _initialized = False

File: runtimes/langgraph-ts/src/otel.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * OpenTelemetry auto-init for the LangGraph (TypeScript) adapter.
 *
 * The router sidecar exposes an OTLP/HTTP collector at
 * `/v1/traces` and `/v1/metrics`. We default to the loopback address
 * so egress-guard can keep the pod sealed; an operator may override
 * via `OTEL_EXPORTER_OTLP_ENDPOINT` if they front the collector
 * elsewhere.
 *
 * `initTelemetry()` is idempotent: a second call is a no-op once a
 * tracer/meter provider is set globally.
 */

const DEFAULT_OTLP_TRACES_ENDPOINT = 'http://127.0.0.1:8443/v1/traces';
const DEFAULT_OTLP_METRICS_ENDPOINT = 'http://127.0.0.1:8443/v1/metrics';

let initialized = false;

function otlpEndpoint(envVar: string, fallback: string): string {
  const raw = process.env[envVar];
  return raw && raw.length > 0 ? raw : fallback;
}

export interface InitTelemetryOptions {
  serviceName: string;
  serviceVersion?: string;
  tracesEndpoint?: string;
  metricsEndpoint?: string;
}

export async function initTelemetry(
  opts: InitTelemetryOptions,
): Promise<void> {
  if (initialized) {
    return;
  }

  // Lazy imports so the module imports cheaply when tests just probe
  // constants, and so a missing OTel dep doesn't crash bootstrap.
  let api:
    | typeof import('@opentelemetry/api')
    | undefined;
  try {
    api = await import('@opentelemetry/api');
    const { Resource } = await import('@opentelemetry/resources');
    const {
      ATTR_SERVICE_NAME,
      ATTR_SERVICE_VERSION,
    } = await import('@opentelemetry/semantic-conventions');
    const { NodeTracerProvider } = await import(
      '@opentelemetry/sdk-trace-node'
    );
    const { BatchSpanProcessor } = await import(
      '@opentelemetry/sdk-trace-base'
    );
    const { OTLPTraceExporter } = await import(
      '@opentelemetry/exporter-trace-otlp-http'
    );
    const {
      MeterProvider,
      PeriodicExportingMetricReader,
    } = await import('@opentelemetry/sdk-metrics');
    const { OTLPMetricExporter } = await import(
      '@opentelemetry/exporter-metrics-otlp-http'
    );

    const tracesUrl =
      opts.tracesEndpoint ??
      otlpEndpoint(
        'OTEL_EXPORTER_OTLP_TRACES_ENDPOINT',
        otlpEndpoint(
          'OTEL_EXPORTER_OTLP_ENDPOINT',
          DEFAULT_OTLP_TRACES_ENDPOINT,
        ),
      );
    const metricsUrl =
      opts.metricsEndpoint ??
      otlpEndpoint(
        'OTEL_EXPORTER_OTLP_METRICS_ENDPOINT',
        DEFAULT_OTLP_METRICS_ENDPOINT,
      );

    const resource = new Resource({
      [ATTR_SERVICE_NAME]: opts.serviceName,
      [ATTR_SERVICE_VERSION]: opts.serviceVersion ?? '0.1.0',
      'service.namespace': 'kars',
      'kars.runtime.kind': 'LangGraph',
      'kars.runtime.language': 'typescript',
    });

    const tracerProvider = new NodeTracerProvider({ resource });
    tracerProvider.addSpanProcessor(
      new BatchSpanProcessor(new OTLPTraceExporter({ url: tracesUrl })),
    );
    tracerProvider.register();

    const meterProvider = new MeterProvider({
      resource,
      readers: [
        new PeriodicExportingMetricReader({
          exporter: new OTLPMetricExporter({ url: metricsUrl }),
        }),
      ],
    });
    api.metrics.setGlobalMeterProvider(meterProvider);

    await instrumentHttp();

    initialized = true;
    // eslint-disable-next-line no-console
    console.error(
      `[kars-runtime-langgraph-ts] OTel initialized: traces=${tracesUrl} metrics=${metricsUrl}`,
    );
  } catch (err) {
    // Telemetry must never crash the agent.
    // eslint-disable-next-line no-console
    console.warn(
      '[kars-runtime-langgraph-ts] initTelemetry failed:',
      err,
    );
  }
}

async function instrumentHttp(): Promise<void> {
  try {
    const { registerInstrumentations } = await import(
      '@opentelemetry/instrumentation'
    );
    const { HttpInstrumentation } = await import(
      '@opentelemetry/instrumentation-http'
    );
    // Node.js 22 / undici fetch coverage is best-effort. The
    // dedicated undici instrumentation, when available, captures the
    // built-in fetch.
    const instrumentations: Array<unknown> = [new HttpInstrumentation()];
    try {
      const undici = await import('@opentelemetry/instrumentation-undici');
      // The export name varies across versions; probe defensively.
      const Cls =
        (undici as { UndiciInstrumentation?: new () => unknown })
          .UndiciInstrumentation ??
        undefined;
      if (Cls) {
        instrumentations.push(new Cls());
      }
    } catch {
      // Optional dep missing — fall back to http instrumentation only.
    }
    registerInstrumentations({
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      instrumentations: instrumentations as any,
    });
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn(
      '[kars-runtime-langgraph-ts] HTTP instrumentation failed:',
      err,
    );
  }
}

/** Test-only hook to reset the module-level init flag. */
export function _resetForTests(): void {
  initialized = false;
}
